README.md
# idkit_tableview

### 1.Introduction

This component supports two kinds of list display, namely single list and grouped list. It supports the view update of a specified element in the list and the view update of a specified group and the view update of the entire list. It does not support the movement and deletion of element views in the table.

### 2. Two types of lists

```dart
enum TableViewStyle {
  plain, // regular table view
  group, // sections are grouped together
}
```

The default is a regular list.

### 3.The control update object of the table view

##### 1. Class Name: IDKitUpdateControl

##### 2. Class definition

```dart
class IDKitUpdateControl {
  /// Factory method to initialize the object.
  factory IDKitUpdateControl() => IDKitUpdateControl._();

  /// Basic initial construction method.
  IDKitUpdateControl._() {
    _updateStreamController = StreamController<IDKitUpdateType>();
  }

  /// Event forwarding subscribers are generated by default.
  StreamController<IDKitUpdateType>? _updateStreamController;
  StreamController<IDKitUpdateType>? get updateStreamController {
    _updateStreamController =
        _updateStreamController ?? IDKitUpdateControl().updateStreamController;
    return _updateStreamController;
  }
  ...... // 省略代码
}
```

##### 3. Update method

```dart
  /// Update the head view of the tableview.
  void updateHeaderTableView()

  /// Update the bottom view of the tableview.
  void updateFotterTableView()

  /// Update the head view of section.
  void updateHeaderSection(int section)

  /// Update the bottom view of section.
  void updateFotterSection(int section)

  /// Update the view of a row in section.
  void updateRowInSection(IDKitIndexPath indexPath)

  /// Update the separate view of section.
  void updateSeparateInSection(int section)

  /// Update all views in a section.
  void updateRowsInSection(int section)

  /// All table views are updated.
  void updateTableView()
```

### 4. Introduction to some functions of this component

```dart
// Set the head view of the table view
1. headerInTableView: (Widget view) -> Widget

// Set the fotter view of the table view
2. footerInTableView: (Widget view) -> Widget

// Set the head view of a group of table views
3. headerInSection: (_, int section) -> Widget

// Set the fotter view of a group of table views
4. footerInSection: (_, int section) -> Widget

// The number of groups in the table view
5. numberOfSection: (_) -> Int

// The number of all elements in each group in the table view
6. numberOfRowInSection: (_, int section) -> Int

// The construction of each group of views containing all elements in the table view
7. itemForRowAtIndexPath: (_, IDKitIndexPath indexPath) -> Widget

// Split view between each group of views in the table view
8. separateInSection: (_, int section) -> Widget
```

### 5. Component example demonstration

##### 1. Single group list instance

```dart
/// Single table view
class SingleTableView extends StatefulWidget {
  const SingleTableView({Key? key}) : super(key: key);

  @override
  _SingleTableViewState createState() => _SingleTableViewState();
}

class _SingleTableViewState extends State<SingleTableView> {
  final List<int> list = List<int>.generate(10, (int index) => 0);
  final IDKitUpdateControl updateControl = IDKitUpdateControl();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Single List'),
      ),
      body: Container(
        alignment: Alignment.center,
        child: IDKitTableView(
          updateControl: updateControl,
          itemForRowAtIndexPath: (_, IDKitIndexPath indexPath) {
            return Container(
              height: 60,
              alignment: Alignment.center,
              color: Colors.red,
              child: Row(
                children: <Widget>[
                  const Icon(Icons.ac_unit_outlined),
                  GestureDetector(
                    onTap: () {
                      list[indexPath.row] += 1;
                      updateControl.updateRowInSection(indexPath);
                    },
                    child: Text('item -- ${indexPath.row}'),
                  ),
                  const Spacer(),
                  Text(list[indexPath.row].toString()),
                ],
              ),
            );
          },
          numberOfRowInSection: (_, __) {
            return list.length;
          },
        ),
      ),
    );
  }
}
```

##### 2. Multi-group list instance

```dart
class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final Map<String, String> map = <String, String>{
    'Num': '110',
    'Count': '220'
  };
  List<int> list = <int>[1, 2, 3, 4, 5, 6];
  final IDKitUpdateControl _updateTableView = IDKitUpdateControl();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Multiple List'),
      ),
      body: Container(
        alignment: Alignment.center,
        child: IDKitTableView(
          style: TableViewStyle.group,
          updateControl: _updateTableView,
          headerInTableView: (Widget view) {
            return Container(
              height: 100,
              color: Colors.pink,
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: <Widget>[
                  GestureDetector(
                    child: Container(
                      padding: const EdgeInsets.all(5),
                      decoration: BoxDecoration(
                        border: Border.all(),
                      ),
                      child: Text(' Single update-${map['Num']}'),
                    ),
                    onTap: () {
                      map['Num'] = '330';
                      _updateTableView.updateHeaderTableView();
                    },
                  ),
                  GestureDetector(
                    child: Container(
                      padding: const EdgeInsets.all(5),
                      decoration: BoxDecoration(
                        border: Border.all(),
                      ),
                      child: Text('Multiple updates-${map['Num']}'),
                    ),
                    onTap: () {
                      map['Num'] = '430';
                      list = list.map((int e) => e * 10).toList();
                      _updateTableView.updateRowsInSection(1);
                    },
                  ),
                  GestureDetector(
                    child: Container(
                      padding: const EdgeInsets.all(5),
                      decoration: BoxDecoration(
                        border: Border.all(),
                      ),
                      child: Text('Update all-${map['Num']}'),
                    ),
                    onTap: () {
                      _updateTableView.updateTableView();
                    },
                  )
                ],
              ),
            );
          },
          footerInTableView: (_) {
            return Container(
              height: 100,
              color: Colors.pink,
              child: Text('Table view bottom - ${map['Num']}'),
            );
          },
          headerInSection: (_, int section) {
            return Container(
              height: 60,
              color: Colors.yellow,
              child: Text(
                  'Table view head view of $section group - ${map['Num']}'),
            );
          },
          footerInSection: (_, int section) {
            return Container(
              height: 60,
              color: Colors.yellow,
              child: Text('Table view fotter view of $section group'),
            );
          },
          numberOfSection: (_) => 4,
          numberOfRowInSection: (_, int section) {
            final int res = section % 2;
            return res == 0 ? 2 : 6;
          },
          itemForRowAtIndexPath: (_, IDKitIndexPath indexPath) {
            if (indexPath.section == 1) {
              return Card(
                child: Text('Card - ${list[indexPath.row]}'),
              );
            } else {
              return Container(
                alignment: Alignment.center,
                child: Row(
                  children: const <Widget>[
                    Icon(Icons.access_alarm_outlined),
                    Text('Item'),
                  ],
                ),
                height: 100,
              );
            }
          },
          hideDivider: false,
          separateInSection: (_, int section) {
            return Container(
              height: 20,
              color: Colors.white,
              child:
                  Text('------------ Group split view -$section ------------'),
            );
          },
        ),
      ),
    );
  }
}
```


CHANGELOG.md
## 0.0.1

### Partial update processing of the table view list

##### 1. There are two types of extended lists

```dart
enum TableViewStyle {
  plain, // regular table view
  group, // sections are grouped together
}
```

##### 2. List extension method

```dart
// Set the head view of the table view
1. headerInTableView: (Widget view) -> Widget

// Set the fotter view of the table view
2. footerInTableView: (Widget view) -> Widget

// Set the head view of a group of table views
3. headerInSection: (_, int section) -> Widget

// Set the fotter view of a group of table views
4. footerInSection: (_, int section) -> Widget

// The number of groups in the table view
5. numberOfSection: (_) -> Int

// The number of all elements in each group in the table view
6. numberOfRowInSection: (_, int section) -> Int

// The construction of each group of views containing all elements in the table view
7. itemForRowAtIndexPath: (_, IDKitIndexPath indexPath) -> Widget

// Split view between each group of views in the table view
8. separateInSection: (_, int section) -> Widget

```

##### 3. List update extension method

```dart
/// Update the head view of the tableview.
void updateHeaderTableView()

/// Update the bottom view of the tableview.
void updateFotterTableView()

/// Update the head view of section.
void updateHeaderSection(int section)

/// Update the bottom view of section.
void updateFotterSection(int section)

/// Update the view of a row in section.
void updateRowInSection(IDKitIndexPathindexPath)

/// Update the separate view of section.
void updateSeparateInSection(int section)

/// Update all views in a section.
void updateRowsInSection(int section)

/// All table views are updated.
void updateTableView()
```

##### 4. For a more detailed introduction, please check [README.md](https://github.com/zhoushuangjian001/idkit_tableview#readme)


test/idkit_tableview_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:idkit_tableview/idkit_tableview.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/src/tableview.dart
import 'dart:async' show StreamController;
import 'dart:collection' show HashMap;
import 'package:flutter/material.dart';
import 'package:idkit_tableview/idkit_tableview.dart';
import 'package:idkit_tableview/src/tableview_definition.dart';
import 'package:idkit_tableview/src/tableview_indexpath.dart';
import 'package:idkit_tableview/src/tableview_stream.dart';
import 'package:idkit_tableview/src/tableview_style.dart';

class IDKitTableView extends StatefulWidget {
  const IDKitTableView({
    Key? key,
    this.controller,
    this.headerInTableView,
    this.footerInTableView,
    this.headerInSection,
    this.footerInSection,
    this.numberOfSection,
    this.numberOfRowInSection,
    this.itemForRowAtIndexPath,
    this.divider,
    this.hideDivider = false,
    this.separateInSection,
    this.updateControl,
    this.style = TableViewStyle.plain,
    this.padding,
    this.margin,
    this.decoration,
    this.color,
    this.width,
    this.height,
    this.physics,
    this.shrinkWrap = false,
    this.primary,
    this.scrollDirection = Axis.vertical,
  }) : super(key: key);

  /// List sliding control object.
  final ScrollController? controller;

  /// The head of the list.
  final WidgetForHeaderInTableView? headerInTableView;

  /// The footer of the list.
  final WidgetForFooterInTableView? footerInTableView;

  /// The head of the section.
  final WidgetForHeaderInSection? headerInSection;

  /// The footer of the section.
  final WidgetForFooterInSection? footerInSection;

  /// The number of inner groups in the list.
  final NumberOfSectionInTableView? numberOfSection;

  /// The number of items in each group.
  final NumberOfRowInSection? numberOfRowInSection;

  /// View construction of list items.
  final ItemForRowAtIndexPath? itemForRowAtIndexPath;

  /// Dividing line between elements.
  final Widget? divider;

  /// Whether to hide the dividing line between elements.
  /// true: hide ; false: Not hidden.
  final bool hideDivider;

  /// Split widget between each section.
  final SeparateInSection? separateInSection;

  /// TableView update control object.
  final IDKitUpdateControl? updateControl;

  /// Type of table view.
  final TableViewStyle style;

  /// The padding of the table view.
  final EdgeInsetsGeometry? padding;

  /// The margins of the table view.
  final EdgeInsetsGeometry? margin;

  /// Rear decoration of table view.
  final Decoration? decoration;

  /// The background color of the table view.
  final Color? color;

  /// The width of the table view.
  final double? width;

  /// The height of the table view.
  final double? height;

  /// The scroll type of the table view.
  final ScrollPhysics? physics;

  /// Table view element display form.
  final bool shrinkWrap;

  /// Whether the table view tracks the parent scroll view scrolling.
  final bool? primary;

  /// The direction in which the table view scrolls.
  final Axis scrollDirection;

  @override
  _IDKitTableViewState createState() => _IDKitTableViewState();
}

class _IDKitTableViewState extends State<IDKitTableView> {
  /// Store all displayed view collections.
  List<Widget> itemsList = <Widget>[];

  /// The key-value collection of the subscribers of each module.
  late HashMap<String, StreamController<bool>> streamControllerMap =
      HashMap<String, StreamController<bool>>();

  /// Prevent scrolling multiple times to build the view.
  late HashMap<String, Widget> widgetsMap = HashMap<String, Widget>();

  /// Reset the collection to avoid duplication.
  void _cleanAllList() {
    itemsList.clear();
    widgetsMap.clear();
    streamControllerMap.clear();
  }

  /// Integrate data into a collection
  void _assembleDataToList() {
    // Process storage collection data.
    _cleanAllList();
    // Top view of table view.
    if (widget.headerInTableView != null) {
      const String key = 'com.idkit.header';
      _addStreamWidgetToList(key, (_, AsyncSnapshot<bool> snapshot) {
        final bool state = snapshot.data!;
        final bool contains = widgetsMap.containsKey(key);
        if (state) {
          final Widget child = widget.headerInTableView!(widget);
          if (contains) {
            widgetsMap.update(key, (_) => child);
          } else {
            widgetsMap.putIfAbsent(key, () => child);
          }
          return child;
        } else {
          if (contains) {
            return widgetsMap[key]!;
          } else {
            final Widget child = widget.headerInTableView!(widget);
            widgetsMap.putIfAbsent(key, () => child);
            return child;
          }
        }
      });
    }

    // View processing for each part of the table view.
    final bool _style = widget.style == TableViewStyle.plain;
    final int totalSection =
        _style ? 1 : (widget.numberOfSection?.call(widget) ?? 1);

    for (int section = 0; section < totalSection; section++) {
      // Header view.
      if (widget.headerInSection != null) {
        final String key = 'com.idkit.section.header-$section';
        _addStreamWidgetToList(key, (_, AsyncSnapshot<bool> snapshot) {
          final bool state = snapshot.data!;
          final bool contains = widgetsMap.containsKey(key);
          if (state) {
            final Widget child = widget.headerInSection!(widget, section);
            if (contains) {
              widgetsMap.update(key, (_) => child);
            } else {
              widgetsMap.putIfAbsent(key, () => child);
            }
            return child;
          } else {
            if (contains) {
              return widgetsMap[key]!;
            } else {
              final Widget child = widget.headerInSection!(widget, section);
              widgetsMap.putIfAbsent(key, () => child);
              return child;
            }
          }
        });
      }
      final int rowCountInSection =
          widget.numberOfRowInSection?.call(widget, section) ?? 0;
      // Row view.
      for (int row = 0; row < rowCountInSection; row++) {
        final IDKitIndexPath indexPath =
            IDKitIndexPath(section: section, row: row);
        final String key = 'com.idkit.section.row-$section-$row';
        _addStreamWidgetToList(key, (_, AsyncSnapshot<bool> snapshot) {
          final bool state = snapshot.data!;
          final bool contains = widgetsMap.containsKey(key);
          if (state) {
            final Widget child =
                widget.itemForRowAtIndexPath?.call(widget, indexPath) ??
                    Container();
            if (contains) {
              widgetsMap.update(key, (_) => child);
            } else {
              widgetsMap.putIfAbsent(key, () => child);
            }
            return child;
          } else {
            if (contains) {
              return widgetsMap[key]!;
            } else {
              final Widget child =
                  widget.itemForRowAtIndexPath?.call(widget, indexPath) ??
                      Container();
              widgetsMap.putIfAbsent(key, () => child);
              return child;
            }
          }
        });
        // Divider line.
        if (!widget.hideDivider) {
          itemsList.add(
            widget.divider ??
                const Divider(color: Colors.green, height: 0.5, thickness: 0.5),
          );
        }
      }
      // Fotter view.
      if (widget.footerInSection != null) {
        final String key = 'com.idkit.section.fotter-$section';
        _addStreamWidgetToList(key, (_, AsyncSnapshot<bool> snapshot) {
          final bool state = snapshot.data!;
          final bool contains = widgetsMap.containsKey(key);
          if (state) {
            final Widget child = widget.footerInSection!(widget, section);
            if (contains) {
              widgetsMap.update(key, (_) => child);
            } else {
              widgetsMap.putIfAbsent(key, () => child);
            }
            return child;
          } else {
            if (contains) {
              return widgetsMap[key]!;
            } else {
              final Widget child = widget.footerInSection!(widget, section);
              widgetsMap.putIfAbsent(key, () => child);
              return child;
            }
          }
        });
      }

      // Separate view in section.
      if (widget.separateInSection != null &&
          section != totalSection - 1 &&
          !_style) {
        final String key = 'com.idkit.separate.section-$section';
        _addStreamWidgetToList(key, (_, AsyncSnapshot<bool> snapshot) {
          final bool state = snapshot.data!;
          final bool contains = widgetsMap.containsKey(key);
          if (state) {
            final Widget child = widget.separateInSection!(widget, section);
            if (contains) {
              widgetsMap.update(key, (_) => child);
            } else {
              widgetsMap.putIfAbsent(key, () => child);
            }
            return child;
          } else {
            if (contains) {
              return widgetsMap[key]!;
            } else {
              final Widget child = widget.separateInSection!(widget, section);
              widgetsMap.putIfAbsent(key, () => child);
              return child;
            }
          }
        });
      }
    }

    // Is it the bottom view of the table view.
    if (widget.footerInTableView != null) {
      const String key = 'com.idkit.fotter';
      _addStreamWidgetToList(key, (_, AsyncSnapshot<bool> snapshot) {
        final bool state = snapshot.data!;
        final bool contains = widgetsMap.containsKey(key);
        if (state) {
          final Widget child = widget.footerInTableView!(widget);
          if (contains) {
            widgetsMap.update(key, (_) => child);
          } else {
            widgetsMap.putIfAbsent(key, () => child);
          }
          return child;
        } else {
          if (contains) {
            return widgetsMap[key]!;
          } else {
            final Widget child = widget.footerInTableView!(widget);
            widgetsMap.putIfAbsent(key, () => child);
            return child;
          }
        }
      });
    }
  }

  /// Add controller components with control stream to the global collection.
  void _addStreamWidgetToList(String key, AsyncWidgetBuilder<bool> builder) {
    final StreamController<bool> streamController =
        StreamController<bool>.broadcast();
    streamControllerMap.putIfAbsent(key, () => streamController);
    final Widget streamWidget =
        StreamWidget(streamController: streamController, builder: builder);
    itemsList.add(streamWidget);
  }

  /// Forward events for updated tableview.
  @override
  void initState() {
    widget.updateControl?.updateStreamController?.stream
        .listen((IDKitUpdateType type) {
      _updateMethod(type);
    });
    super.initState();
  }

  /// Initially build the view.
  @override
  Widget build(BuildContext context) {
    _assembleDataToList();
    return Container(
      padding: widget.padding,
      margin: widget.margin,
      decoration: widget.decoration,
      color: widget.color,
      width: widget.width,
      height: widget.height,
      child: ListView.builder(
        controller: widget.controller,
        physics: widget.physics,
        shrinkWrap: widget.shrinkWrap,
        primary: widget.primary,
        scrollDirection: widget.scrollDirection,
        itemBuilder: (_, int index) => RepaintBoundary(child: itemsList[index]),
        itemCount: itemsList.length,
        cacheExtent: itemsList.length * 0.4,
      ),
    );
  }

  /// Dispose system.
  @override
  void dispose() {
    disposeStreamController();
    super.dispose();
  }

  /// Dispose of all controlled stream with controller.
  void disposeStreamController() {
    streamControllerMap
        .forEach((_, StreamController<bool> item) => item.close());
  }

  /// The general method of updating the view.
  void _updateMethod(IDKitUpdateType updateType) {
    final String key = updateType.key;
    if (key.isNotEmpty) {
      switch (updateType.refresh) {
        case TableViewRefresh.all:
          if (mounted) {
            setState(() {});
          }
          break;
        case TableViewRefresh.section:
          for (final String keyItem in streamControllerMap.keys) {
            if (keyItem.contains(key)) {
              streamControllerMap[keyItem]?.add(true);
            }
          }
          break;
        default:
          streamControllerMap[key]?.add(true);
      }
    }
  }
}


lib/src/tableview_update.dart
import 'dart:async' show StreamController;
import 'package:idkit_tableview/src/tableview_indexpath.dart'
    show IDKitIndexPath;
import 'package:idkit_tableview/src/tableview_style.dart' show TableViewRefresh;

/// TableView update control object class.
class IDKitUpdateControl {
  /// Factory method to initialize the object.
  factory IDKitUpdateControl() => IDKitUpdateControl._();

  /// Basic initial construction method.
  IDKitUpdateControl._() {
    _updateStreamController = StreamController<IDKitUpdateType>();
  }

  /// Event forwarding subscribers are generated by default.
  StreamController<IDKitUpdateType>? _updateStreamController;
  StreamController<IDKitUpdateType>? get updateStreamController {
    _updateStreamController =
        _updateStreamController ?? IDKitUpdateControl().updateStreamController;
    return _updateStreamController;
  }

  /// Update the head view of the tableview.
  void updateHeaderTableView() {
    const String key = 'com.idkit.header';
    _updateStreamController?.add(const IDKitUpdateType(key: key));
  }

  /// Update the bottom view of the tableview.
  void updateFotterTableView() {
    const String key = 'com.idkit.fotter';
    _updateStreamController?.add(const IDKitUpdateType(key: key));
  }

  /// Update the head view of section.
  void updateHeaderSection(int section) {
    final String key = 'com.idkit.section.header-$section';
    _updateStreamController?.add(IDKitUpdateType(key: key));
  }

  /// Update the bottom view of section.
  void updateFotterSection(int section) {
    final String key = 'com.idkit.section.fotter-$section';
    _updateStreamController?.add(IDKitUpdateType(key: key));
  }

  /// Update the view of a row in section.
  void updateRowInSection(IDKitIndexPath indexPath) {
    final String key =
        'com.idkit.section.row-${indexPath.section}-${indexPath.row}';
    _updateStreamController?.add(IDKitUpdateType(key: key));
  }

  /// Update the separate view of section.
  void updateSeparateInSection(int section) {
    final String key = 'com.idkit.separate.section-$section';
    _updateStreamController?.add(IDKitUpdateType(key: key));
  }

  /// Update all views in a section.
  void updateRowsInSection(int section) {
    final String key = 'com.idkit.section.row-$section';
    _updateStreamController
        ?.add(IDKitUpdateType(key: key, refresh: TableViewRefresh.section));
  }

  /// All table views are updated.
  void updateTableView() {
    const String key = 'com.idkit.tableview';
    _updateStreamController
        ?.add(const IDKitUpdateType(key: key, refresh: TableViewRefresh.all));
  }
}

/// The form of the tableview update.
class IDKitUpdateType {
  const IDKitUpdateType({
    required this.key,
    this.refresh = TableViewRefresh.single,
  });

  /// Updated key.
  final String key;

  /// Update form.
  final TableViewRefresh refresh;
}


lib/src/tableview_definition.dart
import 'package:flutter/material.dart' show Widget;
import 'package:idkit_tableview/src/tableview_indexpath.dart';

/// The main method declaration of the list.
/// Get the number of section in tableView.
typedef NumberOfSectionInTableView = int Function(Widget tableView);

/// Get the number of row in section.
typedef NumberOfRowInSection = int Function(Widget tableView, int section);

/// Get the head of the entire list.
typedef WidgetForHeaderInTableView = Widget Function(Widget tableView);

/// Get the head of the entire list.
typedef WidgetForFooterInTableView = Widget Function(Widget tableView);

typedef HeightForRowAtIndexPath = double Function(
    Widget tableView, int section);

/// Get the head of the entire section.
typedef WidgetForHeaderInSection = Widget Function(
    Widget tableView, int section);

/// Get the head of the entire section.
typedef WidgetForFooterInSection = Widget Function(
    Widget tableView, int section);

/// View construction of list items.
typedef ItemForRowAtIndexPath = Widget Function(
    Widget tableView, IDKitIndexPath indexPath);

/// Split widget between each section.
typedef SeparateInSection = Widget Function(Widget tableView, int section);


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:idkit_tableview/idkit_tableview.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final Map<String, String> map = <String, String>{
    'Num': '110',
    'Count': '220'
  };
  List<int> list = <int>[1, 2, 3, 4, 5, 6];
  final IDKitUpdateControl _updateTableView = IDKitUpdateControl();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Multiple List'),
      ),
      body: Container(
        alignment: Alignment.center,
        child: IDKitTableView(
          style: TableViewStyle.group,
          updateControl: _updateTableView,
          headerInTableView: (Widget view) {
            return Container(
              height: 100,
              color: Colors.pink,
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: <Widget>[
                  GestureDetector(
                    child: Container(
                      padding: const EdgeInsets.all(5),
                      decoration: BoxDecoration(
                        border: Border.all(),
                      ),
                      child: Text(' Single update-${map['Num']}'),
                    ),
                    onTap: () {
                      map['Num'] = '330';
                      _updateTableView.updateHeaderTableView();
                    },
                  ),
                  GestureDetector(
                    child: Container(
                      padding: const EdgeInsets.all(5),
                      decoration: BoxDecoration(
                        border: Border.all(),
                      ),
                      child: Text('Multiple updates-${map['Num']}'),
                    ),
                    onTap: () {
                      map['Num'] = '430';
                      list = list.map((int e) => e * 10).toList();
                      _updateTableView.updateRowsInSection(1);
                    },
                  ),
                  GestureDetector(
                    child: Container(
                      padding: const EdgeInsets.all(5),
                      decoration: BoxDecoration(
                        border: Border.all(),
                      ),
                      child: Text('Update all-${map['Num']}'),
                    ),
                    onTap: () {
                      _updateTableView.updateTableView();
                    },
                  )
                ],
              ),
            );
          },
          footerInTableView: (_) {
            return Container(
              height: 100,
              color: Colors.pink,
              child: Text('Table view bottom - ${map['Num']}'),
            );
          },
          headerInSection: (_, int section) {
            return Container(
              height: 60,
              color: Colors.yellow,
              child: Text(
                  'Table view head view of $section group - ${map['Num']}'),
            );
          },
          footerInSection: (_, int section) {
            return Container(
              height: 60,
              color: Colors.yellow,
              child: Text('Table view fotter view of $section group'),
            );
          },
          numberOfSection: (_) => 4,
          numberOfRowInSection: (_, int section) {
            final int res = section % 2;
            return res == 0 ? 2 : 6;
          },
          itemForRowAtIndexPath: (_, IDKitIndexPath indexPath) {
            if (indexPath.section == 1) {
              return Card(
                child: Text('Card - ${list[indexPath.row]}'),
              );
            } else {
              return Container(
                alignment: Alignment.center,
                child: Row(
                  children: const <Widget>[
                    Icon(Icons.access_alarm_outlined),
                    Text('Item'),
                  ],
                ),
                height: 100,
              );
            }
          },
          hideDivider: false,
          separateInSection: (_, int section) {
            return Container(
              height: 20,
              color: Colors.white,
              child:
                  Text('------------ Group split view -$section ------------'),
            );
          },
        ),
      ),
    );
  }
}

/// Single table view
class SingleTableView extends StatefulWidget {
  const SingleTableView({Key? key}) : super(key: key);

  @override
  _SingleTableViewState createState() => _SingleTableViewState();
}

class _SingleTableViewState extends State<SingleTableView> {
  final List<int> list = List<int>.generate(10, (int index) => 0);
  final IDKitUpdateControl updateControl = IDKitUpdateControl();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Single List'),
      ),
      body: Container(
        alignment: Alignment.center,
        child: IDKitTableView(
          updateControl: updateControl,
          itemForRowAtIndexPath: (_, IDKitIndexPath indexPath) {
            return Container(
              height: 60,
              alignment: Alignment.center,
              color: Colors.red,
              child: Row(
                children: <Widget>[
                  const Icon(Icons.ac_unit_outlined),
                  GestureDetector(
                    onTap: () {
                      list[indexPath.row] += 1;
                      updateControl.updateRowInSection(indexPath);
                    },
                    child: Text('item -- ${indexPath.row}'),
                  ),
                  const Spacer(),
                  Text(list[indexPath.row].toString()),
                ],
              ),
            );
          },
          numberOfRowInSection: (_, __) {
            return list.length;
          },
        ),
      ),
    );
  }
}
